x/incentive/keeper/adapters/hard_borrow/adapter_test.go

```diff
@@ -0,0 +1,285 @@
+					"ukava",
+					"usd",
+					nil,
+					true,
+				),
+			},
+		),
+		[]pricefeedtypes.PostedPrice{
+			pricefeedtypes.NewPostedPrice(
+				"ukava:usd:30",
+				suite.addrs[0],
+				sdk.MustNewDecFromStr("1.5"),
+				suite.ctx.BlockTime().Add(time.Hour),
+			),
+		},
+	)
+
+	cdc := suite.app.AppCodec()
+	suite.app.InitializeFromGenesisStates(
+		app.GenesisState{
+			hardtypes.ModuleName:      cdc.MustMarshalJSON(&hardGs),
+			pricefeedtypes.ModuleName: cdc.MustMarshalJSON(&pricefeedGs),
+		},
+	)
+}
+
+func (suite *HardBorrowAdapterTestSuite) TestHardAdapter_OwnerSharesBySource() {
+	hardKeeper := suite.app.GetHardKeeper()
+
+	// Need some deposits in order to borrow
+	err := hardKeeper.Deposit(
+		suite.ctx,
+		suite.addrs[0],
+		sdk.NewCoins(
+			sdk.NewCoin(suite.denomA, sdk.NewInt(100000)),
+		),
+	)
+	suite.NoError(err)
+
+	err = hardKeeper.Deposit(
+		suite.ctx,
+		suite.addrs[1],
+		sdk.NewCoins(
+			sdk.NewCoin(suite.denomA, sdk.NewInt(100000)),
+		),
+	)
+	suite.NoError(err)
+
+	// Actual borrows now
+	err = hardKeeper.Borrow(
+		suite.ctx,
+		suite.addrs[0],
+		sdk.NewCoins(
+			sdk.NewCoin(suite.denomA, sdk.NewInt(100)),
+		),
+	)
+	suite.NoError(err)
+
+	err = hardKeeper.Borrow(
+		suite.ctx,
+		suite.addrs[1],
+		sdk.NewCoins(
+			sdk.NewCoin(suite.denomA, sdk.NewInt(250)),
+		),
+	)
+	suite.NoError(err)
+
+	adapter := hard_borrow.NewSourceAdapter(suite.app.GetHardKeeper())
+
+	tests := []struct {
+		name          string
+		giveOwner     sdk.AccAddress
+		giveSourceIDs []string
+		wantShares    map[string]sdk.Dec
+	}{
+		{
+			"depositor has shares",
+			suite.addrs[0],
+			[]string{
+				suite.denomA,
+			},
+			map[string]sdk.Dec{
+				suite.denomA: sdk.NewDecWithPrec(100, 0),
+			},
+		},
+		{
+			"depositor has shares - including empty deposits",
+			suite.addrs[1],
+			[]string{
+				suite.denomA,
+				"unknown",
+			},
+			map[string]sdk.Dec{
+				suite.denomA: sdk.NewDecWithPrec(250, 0),
+				"unknown":    sdk.ZeroDec(),
+			},
+		},
+		{
+			"non-depositor has zero shares",
+			suite.addrs[2],
+			[]string{
+				suite.denomA,
+			},
+			map[string]sdk.Dec{
+				suite.denomA: sdk.ZeroDec(),
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		suite.Run(tt.name, func() {
+			shares := adapter.OwnerSharesBySource(suite.ctx, tt.giveOwner, tt.giveSourceIDs)
+
+			suite.Equal(tt.wantShares, shares)
+		})
+	}
+}
+
+func (suite *HardBorrowAdapterTestSuite) TestHardAdapter_TotalSharesBySource() {
+	hardKeeper := suite.app.GetHardKeeper()
+
+	err := hardKeeper.Deposit(
+		suite.ctx,
+		suite.addrs[0],
+		sdk.NewCoins(
+			sdk.NewCoin(suite.denomA, sdk.NewInt(100000)),
+		),
+	)
+	suite.NoError(err)
+
+	err = hardKeeper.Deposit(
+		suite.ctx,
+		suite.addrs[1],
+		sdk.NewCoins(
+			sdk.NewCoin(suite.denomA, sdk.NewInt(100000)),
+		),
+	)
+	suite.NoError(err)
+
+	err = hardKeeper.Borrow(
+		suite.ctx,
+		suite.addrs[0],
+		sdk.NewCoins(
+			sdk.NewCoin(suite.denomA, sdk.NewInt(100)),
+		),
+	)
+	suite.NoError(err)
+
```

Comment:
Similar to the owner shares test, I think we should test the case where `GetBorrowedCoins` differs from the total source share amount.

Should be able to tigger by:
- borrowing
- stepping through a few blocks to accumulate interest

At that point `GetBorrowedCoins` will include interest, but the `TotalSharesBySource` amount shouldn't.